### DISCUSSION TITLE
Best way to add a PyPI badge/link in the header?

### DISCUSSION THREAD

**Discussion by bluss** [squidfunk/mkdocs-material#7429] [Q&A] [answered]
Hi, I've got this admittedly very basic but very effective and easy site setup for a python package.

What would be the best way to get a link to the [PyPi](https://pypi.org/) entry for my package in the header? I'm thinking it would be placed between the Search field and the git icon in the top right. I think I can find or make a badge that includes the current version number. Is there an existing plugin for this? Is the right thing to override the header partial (copy the whole existing code there)?

https://bluss.github.io/pyproject-local-kernel/

**Comment by kamilkrzyskow** [comment, marked as answer]
Hello @bluss,
I checked the [catalog](https://raw.githubusercontent.com/mkdocs/catalog/main/README.md) and there is a project for badges in markdown in the content:
- https://github.com/six-two/mkdocs-badges

However, as you want to add the badge outside of the content this won't help you, so you're correct in using template customization, and yes you only have to copy the original contents and make your adjustments.
- https://squidfunk.github.io/mkdocs-material/customization/#extending-the-theme
- https://github.com/squidfunk/mkdocs-material/blob/master/src/templates/partials/header.html
- https://github.com/squidfunk/mkdocs-material/blob/master/material/templates/partials/header.html

You can use shields.io to generate the badge based on the URL.
- https://squidfunk.github.io/mkdocs-material/blog/2023/11/30/adding-a-badge-to-your-project/
- Using the `for-the-badge` style makes it flat and more fitting for Material 

You can make a [hook](https://www.mkdocs.org/user-guide/configuration/#hooks) to fetch the latest version of your PyPi project via:
- Parsing XML from the latest release from the RSS feed https://pypi.org/rss/project/pyproject-local-kernel/releases.xml
- Parsing JSON from the API endpoint https://pypi.org/pypi/pyproject-local-kernel/json

You can use [extra](https://www.mkdocs.org/user-guide/configuration/#extra) to pass info from the hook to the templates.

**Reply by bluss** [reply]
Thank you very much for the detailed answer!

**Reply by kamilkrzyskow** [reply]
@bluss 
I forgot to mention, that you should implement some caching for fetched results.
`mkdocs serve` will rerun all of your hook events so your event needs to be aware of already fetched data to not refetch it from PyPi on each restart.

Just save the fetched file in the .cache directory of your project, like the Social cards plugin does it, with a datetime name for the file and fetch it periodically.
